### DISCUSSION TITLE
Neighboring Files: Related Files/Folders (Name Matching)

### DISCUSSION THREAD

**Discussion by Mikenux** [flatpak/xdg-desktop-portal#1259] [Ideas]
## Goal
Allow apps to request access to neighboring files/folders with the same name as a first open file/folder, limited to relevant relationships between file/folder types.

### App Developers
- Use of the file chooser portal.
- A reference file where to define relationships between file/folder types. See "Relationship Rules" below for how to define them.
- Be able to request, in xdg-desktop-portal, the addition of new file types and categories, as well as file types to a category. If allowed, each request must include the requested types (with relevant extensions) and relevant categories, how these relate to each other (between file types, between file types and categories), and a flatpak app to test. The request must be compatible with the relational rules.

### File Types and Categories Database

- Database of file types classified into categories. A specific file type can be a category if it is the only representative of it.
- Limited to file types where files are expected to have the same name and be related to each other. This does not concern folders.
- This database is important because it will be used to simplify and validate the relationships defined in the app reference file. For example, all video types and subtitle types are categorized as "video" and "subtitles", so an access request will request access to all types of subtitles when a video (regardless of its type) is open. It can also be used for portal user interfaces.

### Flatpak Validation

Flatpak must be able to validate relationships defined in the app reference file, including categories and types/extensions under each category, using the database and the relationship rules. When an app is installed, flatpak performs validation and only valid relationships are remembered. When an app is updated, flatpak validates the relationships and compares them to previous ones, spotting new, stable, and removed relationships. New relationships are marked as new. Stable relationships are marked as stable and their set access permissions remain those chosen by the user. Removed relationships have their access permission set removed; these can be marked or not.

### Portal
- Ability to perform the checks listed in "Functioning, Checks and their values" below.
- Ability to select files based on check values.
- Ability to remove access to files or folders selected by the portal (see "Functioning" below). This means that the app can no longer automatically access these the next time the app is launched.
- Ability to detect changes in relationships at app startup (using flatpak validation results).

### Relationships Between Files

For files, the relationships between categories are written (e.g. "video->subtitles"), then the relevant extensions are written for each category (e.g. video: mp4, mkv, webm). The relationships must respect certain rules, depending on the type of file extension.

For extensions with letters (e.g. .webm) and extensions with letters and fixed digits (e.g. .mp4), the rules are as follows:

1. Relationships are unidirectional. For example, if the relationship "video->subtitles" is written because it is wanted, then the relationship "subtitles->video" is not allowed.
2. File categories in a relationship are different. For example, the relationship "video->video" is not allowed.
3. All written extensions belong to the correct category. File extensions under the wrong category are not allowed.

For variable digit extensions (usually for split files), the rules are:

1. The entire extension has at least one letter. Each letter of the entire extension is written, and each digit is replaced, for example, with an asterisk ("*"). Examples: .r01, .r02, etc. become .r**, .part01.rar, .part02.rar, etc. become .part**.rar, and .7z.001, .7z.002, etc. become .7z.***.
2. Relationships can be bidirectional. Examples: bi: .rar->.r**, bi: part**.rar->.part**.rar.
3. Extensions belong to the same file category.

### Relationships Between Folders

For folders, the relationships between folder extensions are written. This means that additional folders, and therefore their contents, can only be opened if a first folder has been opened.

There is only one rule to follow: folder extensions must be different.

## Functioning

Note: The following text is for understanding purposes only. Things can be done differently if wanted; only the end results are important.

### Set Access Permission at App Startup

When the app is launched, whether a file or folder is open, an access permission dialog box is first presented, telling the user what "related" files and folders are, depending on valid relationships defined by the app. This dialog box offers to set access permission for each file category (e.g. subtitles, audio) or for a folder type of a relationship (e.g. video->audio, video-> subtitles, inform->materials). If enabled, access is automatic access to file/folder types based on valid relationships. Alternatively, access permission may affect all relationships at once rather than separately; in this case, a "confirm" permission (i.e. always ask to grant access) can be included.

If there are changes in relationships, the portal must detect them when the app starts. If there are new relationships, a dialog box to set access permission for them is displayed; this dialog can also indicate which relationships were removed if this option was chosen in flatpak validation. Note that changes are handled for relationships between categories, not between specific file types/extensions; informing the user that a specific file type is no longer supported is up to the app developer.

### Access Related Files/Folder

Checks and their values:
- (1) The open item in the file chooser or file manager is a file or folder (i.e. not multiple). Values: "file", "folder".
- (2a) file: There are files of valid types in the same folder as the open file. To do this, the portal refers to file extensions with valid relationships and checks for the existence of files with these extensions. Values: "yes", "no".
- (2b) folder: There is a folder of valid type in the same folder as the open folder. To do this, the portal refers to the folder extension of the valid relationship and checks the existence of a folder with this extension. Values: "yes", "no".
- (3) There are files or folder with the exact same name. The name starts with the first letter or digit and ends before the file/folder extension. Each character in the name, given its case (i.e. uppercase, lowercase), must match. Values: "yes", "no".
- (4) Value of the access permission for each relationship. Values: "never" if set to never grant access, and "automatic" if set to grant access automatically. If setting a global permission for all relationships at once has been chosen, the values are the same but applied globally, and there may also be the value "confirm" to always ask to confirm access to related files/folder.
- Note: For file/folder extensions, consider whether the filesystem is case sensitive. In such a filesystem, an extension written in lowercase (for example mkv) is not the same as the same extension written in uppercase (for example MKV); these must be considered identical.

To begin, the user selects a file or folder with an extension to open from the file chooser or file manager. Then, the portal does check (1). If check (1) is "file", then the portal performs check (2a); if the value is "folder", the portal does check (2b). If check (2a) or (2b) is "no", then the portal stops the checks and only the file or folder selected by the user is opened. However, if they are "yes", the portal selects the relevant files or folders and then continues with check (3).

If check (3) is "no", the portal only opens the file or folder selected by the user. If check (3) is "yes", then the portal selects the files or the folder with the same name as the file or folder selected by the user and performs check (4).

After performing check (4), the portal selects only the files or folder for which "automatic" access has been set, and these are accessible by the app, including the file/folder selected by the user. If all relationships have "never" access permission, then only the file/folder selected by the user is accessible by the app.

However, if a global permission has been chosen, the portal will have one of the following behavior depending on the value of check (4):
- "never": Only the file/folder selected by the user is accessible by the app.
- "confirm": The user must confirm whether they agrees to grant access to the files or folder selected by the portal.
- "automatic": Access to files or folder selected by the portal is automatically granted.

Once the app is no longer running when it has been closed, the portal removes access to the files or the folder selected by the portal.

## A Simpler Portal

<details>
  <summary>View</summary>

#### Portal
- Ability to search for files/folder with the exact same name and in the same folder as a file/folder opened via the file chooser or file manager. The name starts with the first letter or digit and ends before a file/folder extension. Each character in the name, given its case (i.e. uppercase, lowercase), must match. The portal limits name matching to extensions of 1 to 4 characters after a dot and exceptions.
- Ability to perform the checks listed in "Functioning, Checks and their values" below.
- Ability to select files based on check values.
- Ability to remove access to files or folders selected by the portal (see "Functioning" below). This means that the app can no longer automatically access these the next time the app is launched.

#### App Developers
- Use of the file chooser portal.
- They should be allowed to request the addition of extensions that have more than 4 characters after a dot (e.g. .vol000+01.PAR2, .mpv.edl) or that do not have a dot (e.g. -wal for SQlite), since full file and folder name matching will be used. The limitation to extensions of 4 characters maximum is justified by the following example: "Name.S01.E01.mkv" is not the same as "Name.S01.E02.mkv" nor "Name.S02.E01.mkv", which should not match. The request must include the relevant extensions, their types, a file to test, and a sandbox or sandboxable app to test.

### Functioning

Note: The following text is for understanding purposes only. Things can be done differently if wanted; only the end results are important.

Checks and their values:
- (1) The selected item in the file chooser or file manager is a file or folder (i.e. not multiple). Values: "file", "folder".
- (2a) file: There are files with the same file name and in the same folder as the open file. See "Requirements, Portal" for what to consider as a name. Values: "yes", "no".
- (2b) folder: There are other folders with an extension, of the same name, and in the same folder as the open folder. See "Requirements, Portal" for what to consider as a name. Values: "yes", "no".
- (3) The files or folders selected by the portal have different extensions between them and compared to the file or folder selected by the user. Values: "none", "some".
- (4) Value of the access permission. Values: "none" if permission is not set, "never" if set to never grant access, "confirm" if set to confirm access (this one is optional), and "automatic" if set to grant access automatically.
- Note: For file/folder extensions, consider whether the filesystem is case sensitive. In such a filesystem, an extension written in lowercase (for example mkv) is not the same as the same extension written in uppercase (for example MKV); these must be considered identical.

To begin, the user selects a file or folder with an extension to open from the file chooser or file manager. Then, the portal does check (1). If check (1) is "file", then the portal performs check (2a); if the value is "folder", the portal does check (2b).

If check (2a) or (2b) is "no", only the file or folder selected by the user is opened. However, if check (2a) or (2b) is "yes", the portal selects all files or folders having their name matching the name of the file or folder selected by the user, and performs check (3).

If check (3) is "none", then the portal only opens the file or folder selected by the user. If its value is "some", the portal selects all files or folders that have a different extension between them and compared to that of the file or folder selected by the user, and performs check (4).

The portal has one of the following behaviors, depending on the value of check (4):
- "none": A dialog box asking to define the access permission is displayed. The user has the choice of giving access to all files selected by the user and the portal automatically ("automatic"), by confirming access ("confirm"; this is optional), or never ("never"). When the choice is made, the value of check (4) is replaced by the value of the selected choice, and the portal acts according to this new value.
- "never": Access to selected files or folders is not granted.
- "confirm": The user must confirm whether they agrees to grant access to the selected files or folders.
- "automatic": Access to selected files or folders is automatically granted.

Once the app is no longer running when it has been closed, the portal removes access to the files or folders selected by the portal.

</details>

## Questions
- Can access permissions set at app startup can be directly applied, especially for a file or folder opened from the file manager? Or should the app need to be restarted? Or should the dialog to set access be presented before the app starts?

**Comment by ptomato** [comment]
For my use case, I'd need to be able to _create_ the related folder if it didn't exist in (2b). Would that be possible here?

**Reply by Mikenux** [reply]
No, but I had opened https://github.com/flatpak/xdg-desktop-portal/issues/997 for this. This solution is more general and independent of any solution for opening neighboring files. Maybe I can take another look, especially since no new cases have surfaced.

**Comment by Mikenux** [comment]
**Thread for updates to the proposal**

Changes made to the proposal can be viewed using the "edited" button.

Update 1:
- Clarification: A file type can be a category if it is its only representative.
- Clarification: Changes in relationships concern changes between categories, not between specific file types/extensions.

**Comment by J5lx** [comment]


**Reply by Mikenux** [reply]
Thank you for your interest in this proposal :slightly_smiling_face:

First of all, I'm not a developer and this proposal (like others I've opened) needs developer(s).

To answer your questions :

> You mention extensions without a dot, does this include folders as well? In my case I have a project file called basename.ext and an associated project folder called basename_data/ which I need to access. Would that be covered by this proposal?

I didn't know about this case because generally project files require the app to request the opening of several other files. But there is no problem adding this case to the proposal. The thing is that we have a relationship between extensions or what we can assimilate to extensions, so no problem to have `_data` assimilated to an extension.

That said, as another idea, it would be better if all app developers use a common "extension". For example, `.file_extension.d` would be much better because it could become a rule without any definition for app developers because in this form a user can tell that a file and a folder are indeed related.

> it’d be great if an app could declare that access to the related file/folder is a required and thereby cause the portal to grant access on an all-or-nothing basis instead (and, perhaps even more importantly, appropriately communicate this to the user).

No, sorry. It is up to the user to assess whether the request is relevant, and putting a reason in the portal to require access to other files is not a good idea. The fact is that any app can request a folder with an "extension", and the portal is not there to influence the user's decision. If the portal doesn't return the selected file, you can tell the user, and they can change the permission if they initially denied the request. And if the user is not satisfied, then they can attribute any problem to the app and not to the portal. However, the user will have given the project file to the app, but this is better than having the portal require giving all files or nothing. (I hope this is understandable to you)

> There is mention of numbered extensions, but what about files where number precedes the extension, i.e. basename001.ext, basename002.ext, basename003.ext…? This isn’t uncommon for image sequences in video editing.

I suppose we can also equate this to an extension since the number is at the end of the name. What is important is that a request is made to properly describe the case so that it can be assigned to a file type and category that is, as much as possible, meaningful to a user. Worst case scenario, if someone objects, just use a dot before the numbers and tell the user to rename accordingly.

**Reply by J5lx** [reply]
Thank you for your reply, this answers my two questions.

> If the portal doesn't return the selected file, you can tell the user, and they can change the permission if they initially denied the request.

I find this to be somewhat cumbersome, but I suppose I will just make do with it then. I just don’t view “maybe your project is corrupted, maybe its a permission issue, please figure it out yourself” as a particularly great type of error message, which is why I wanted to let the portal take care of the permission aspect. But as long as it’s a deliberate decision and not an oversight I’m good.

**Reply by Mikenux** [reply]
What is needed here is a signal indicating whether the folder exists. However, I think we cannot do this. Only contributors and maintainers of xdg-desktop-portal who are familiar with the portals can say anything about this with certainty.

**Reply by Mikenux** [reply]
Maybe having such a signal is a privacy issue.

However, if we are sure that a specific folder is always required for a specific file or folder, then assuming that they do indeed form a "project" - so that they conceptually represent a single object as you assume - then we can either not return all files/folders or create the partner folder if it does not exist. Then, if no file or folder is returned, we return an error to the user. It would be a matter of making it a general rule, hoping that there are no cases that contradict it.

@ptomato: What do you think of this?

**Reply by ptomato** [reply]
In my use case (two side-by-side folders called "My Project.inform" and "My Project.materials") the .materials folder is allowed to not exist, and will be created if needed. It's possible that a project doesn't have materials, and in that case the folder isn't needed. It would be OK with me if the portal always created the folder, although maybe slightly odd in the case of projects without materials. If the portal denied access to the .inform folder when the .materials folder couldn't be accessed, that would not be so great for my use case.

**Reply by Mikenux** [reply]
@ptomato: When I click "Compile and run the story" (without having created one), it automatically creates the `.materials` folder. So I guess the `.materials` folder is needed for this feature which the user should definitely use. So if access is set to "never", we must logically not give access to any `.materials` folder even if it is selected via the file chooser. In this case I think it makes sense not to give the `.inform` folder to the app.

**Reply by ptomato** [reply]
Not sure I agree with that. You could be trying to browse an Inform project without materials that is on read-only media, in which case you might want to re-save it somewhere else where you can give the app access to create the materials folder. Or you could have accidentally created a non-folder file called "My Project.materials", and I don't think that should prevent the whole project from being opened; for one thing, because it would be very difficult to communicate to the user what the problem was.

**Reply by Mikenux** [reply]
I am thinking of the following:
- Inform users that a file/folder and its related folder are a single object, with no permissions to set.
- The first file or folder can be given to the app if the user wants to open it, whether the related folder exists or not.
- The app assumes that the associated folder does not exist if it was not returned by the portal.
- If the application needs to create the associated folder, possibly with other files inside, then it is https://github.com/flatpak/xdg-desktop-portal/issues/997.
- May have a standardized suffix or extension for the related folder. This would contain at least the extension of the first file or folder. This would help reinforce that two objects are indeed related and avoid possible naming issues. It is preferable to have a suffix for DEs that do not display the default extension. Having something completely standardized allows to have a general rule too. This would involve the user renaming some folders, but it is communicable and does not require changing the way their files are organized.

What are you thinking @ptomato and @J5lx? Thanks!

**Comment by Mikenux** [comment]
For relationships between files, I would like to make it possible to search for related files that are not just neighbors of the first open file. This was requested on the issue of neighboring files. This is important when the user has, for example, video subtitles in a Subtitles subfolder.

However, my concern is performance. So, which is the best solution between the following two?
- Search for related files in subfolders. With this, the portal must search all subfolders before returning the result. The app will not be able to add a folder containing related files as a library.
- Allow the user to define which folders to search. With this, the portal will search only in folders specified by the user. When a folder is defined, the app will not be able to access them by asking to add them as a library.

**Reply by Mikenux** [reply]
Note that the solution where the portal searches for related files in the same folder and in the first subfolders (without going deeper in the hierarchy) is better for the user. This would not force them to think about adding new folders or deleting folders if they renames them or moves related files to a newly created folder. However, performance must be there.

**Comment by Mikenux** [comment]
I would like to rewrite this proposal, as I did for "Neighboring Images" which became "Neighbors Of The Same File Category".

I would also like the case to be identified rather than having the permissions set at app startup. This case identification is done with the file selected by the user and the presence of other files of the same name and relevant types.

For this, I really need to know:
- If the search for related files in the first subfolders is sufficiently performant. This is to be sure that files can be quickly returned to the app, especially when no related files are found.
- If we should treat split files as one big file (related files are always accessed) or as files that can be opened independently. The fact is that having permission to never access associated files is not very relevant because it means not using the app to have a complete file (represented by all its parts). However, split files can usually be opened independently.

**Comment by Be-ing** [comment]
Could this satisfy the use case of reading files that are referenced from a file that the user picked via the file picker, for example importing a playlist into a music application? https://github.com/flatpak/xdg-desktop-portal/discussions/1288

**Comment by Mikenux** [comment]
Closing in favor of https://github.com/flatpak/xdg-desktop-portal/discussions/1290.
